crates/sgx_validation/src/config.rs

```diff
@@ -0,0 +1,130 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+use cryptography::digest::Digest;
+use serde::{Deserialize, Deserializer, Serialize};
+use sgx::parameters::Features;
+
+#[derive(Clone, Deserialize, Debug, Serialize)]
+pub enum SgxFeatures {
+    CET,
+    Debug,
+    EIntKey,
+    KSS,
+    ProvisioningKey,
+}
+
+#[derive(Clone, Deserialize, Debug, Default, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct ConfigIntermediate {
+    /// Values for `mrsigner` in the report body.
+    /// This is the list of public keys which have signed the Enarx binary.
+    #[serde(default)]
+    pub enarx_signer: Vec<Digest>,
+
+    /// Values for `mrenclave` in the report body.
+    /// This is the hash of the Enclave environment after the Enarx binary is loaded
+    /// but before any workload is loaded, so this is a hash of the Enarx binary
+    /// in memory.
+    #[serde(default)]
+    pub enarx_hash: Vec<Digest>,
+
+    /// Values for `features`.
+    #[serde(default)]
+    #[serde(deserialize_with = "from_features")]
+    pub features: u64,
+
+    /// Minimum value for `isv_svn`.
+    pub enclave_security_version: Option<u16>,
+
+    /// Value for `isv_prodid`, do not allow versions below this.
+    pub enclave_product_id: Option<u16>,
+}
+
+fn from_features<'de, D>(deserializer: D) -> Result<u64, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: Vec<SgxFeatures> = Deserialize::deserialize(deserializer)?;
+
+    let mut flags = Features::empty();
+
+    flags |= Features::INIT; // Must be set
+    flags |= Features::MODE64BIT; // Required by Enarx
+
+    for flag in s {
+        match flag {
+            SgxFeatures::CET => {
+                flags |= Features::CET;
+            }
+            SgxFeatures::Debug => {
+                flags |= Features::DEBUG;
+            }
+            SgxFeatures::EIntKey => {
+                flags |= Features::EINIT_KEY;
+            }
+            SgxFeatures::KSS => {
+                flags |= Features::KSS;
+            }
+            SgxFeatures::ProvisioningKey => {
+                flags |= Features::PROVISIONING_KEY;
+            }
+        }
+    }
+
+    Ok(flags.bits())
+}
+
+#[derive(Clone, Debug, Default, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct Config {
+    #[serde(flatten)]
+    pub config: ConfigIntermediate,
```

Comment:
There should not be an intermediate step. It's one step: TOML -> valid `Config` ready-to-use

Reply:
How else do you instruct `serde` to say that one or two fields must be present? From chatting with @haraldh, we came to the conclusion that:

> Key whitelist or enarx digest whitelist, require at least one.

Having an intermediate step seemed to be the only way for `serde` to do this. Is there's something better, please advise.

Reply:
Please schedule a call and we can go through the details, I don't understand the problem

Reply:
Fixed.